### ISSUE TITLE
Docs suggestion: use "prepare" instead of "postinstall"

### ISSUE THREAD

**Issue by nathansmith** [typicode/husky#884]
### TL;DR = Suggestion for the documentation site

This is not a bug "issue," but I wanted to make a suggestion for the docs.

I think this could alleviate some confusion around `postinstall` and remove the necessity for `pinst` entirely.

Currently, the Husky v5 documentation reads like this.

> To automatically have Git hooks enabled after install, edit package.json
>
> ```json
> // package.json
> {
>   "private": true,
>   "scripts": {
>     "postinstall": "husky install"
>   }
> }
> ```

Instead, that could be changed to this. Swap out `posinstall` for `prepare`.

Also, `private` does not matter because `prepare` runs the same regardless.

> To automatically have Git hooks enabled after install, edit package.json
>
> ```json
> // package.json
> {
>   "scripts": {
>     "prepare": "husky install"
>   }
> }
> ```

---

### Longer explanation

I did a writeup for my coworkers, to explain our automated setup for configuring Husky v5.

Basically, this all runs via NPM with the `prepare` command and puts everything into place.

If this is helpful for the docs, feel free to repurpose it.

> This `prepare.sh` file runs after `npm install`, when pulling down dependencies.
>
> ```json
> {
>   "scripts": {
>     "prepare": "./scripts/prepare.sh"
>   }
> }
> ```
> 
> Unlike `postinstall`, the `prepare` command does not run when a NPM package is installed as a dependency in another project.
> 
> Meaning, it will _**not**_ run within a package when someone types this…
> 
> `npm install your-npm-package`
> 
> …to install it for their own project.
>
> But it _**will**_ run when working on `your-npm-package` and installing other dependencies locally. This replicates the behavior we had before with Husky v4.
> 
> Code within the `prepare.sh` file looks like this.
> 
> ```bash
> if [ ! -f './.husky/_/husky.sh' ]; then
>   npx husky install
> fi
> 
> if [ ! -f './.husky/pre-commit' ]; then
>   npx husky add './.husky/pre-commit' 'npx lint-staged'
> fi
> 
> ```
> 
> Code within the conditional logic will create `.husky/pre-commit` if it does not yet exist.
> 
> That adds a snippet (instructions for Husky) at the top of the file, and then appends the actual text that we passed in.
> 
> ```bash
> #!/bin/sh
> . "$(dirname "$0")/_/husky.sh"
> 
> npx lint-staged
> ```

**Comment by blimmer** [comment]
If this change is made to the docs, you'll probably also want to update the migration helper: https://github.com/typicode/husky-4-to-5/blob/30792c918d08c0ebb99e1ef8942d13be91b9dd3e/src/pkg.ts#L65

**Comment by typicode** [comment]
@nathansmith Looks great, thanks a lot for the suggestion and detailed explanation! :+1:

**Comment by typicode** [comment]
I've updated docs, it's a better approach for `npm/pnpm/yarn v1` :+1:
However it seems like Yarn v2 doesn't support `prepare` https://yarnpkg.com/advanced/lifecycle-scripts

**Comment by typicode** [comment]
I've added a dedicated section to the docs for Yarn v2.

Also `husky init` will detect Yarn v2 and use `postinstall [+ pinst]` instead of `prepare`.

**Comment by andresmatasuarez** [comment]
There's an important caveat to the use of `prepare` (or `postinstall` for that matter) that's worth mentioning:

If you have installed `husky` as a devDependency (as is my case and is suggested in the [official docs](https://typicode.github.io/husky/#/?id=install)) and are running `npm publish` as part of a CI or release process or whatever, then it'd simply not work due to `husky not found` error.

I wouldn't want to resort to simply installing `husky` in the regular dependencies and I know there's a whole [official section](https://typicode.github.io/husky/#/?id=disable-husky-in-cidocker) explaining ways to disable `husky` for specific environments, but I wanted to make people coming across this issue in the future aware of this finding

**Comment by david10sing** [comment]
I have package A (components library) where I use husky with prepare and I have another repository (Package B) where I use package A (from private git repository) as well as husky with prepare.

During yarn install, it looks like yarn tries to install package A because it has a prepare script and also install other deps from package B. It appears both installs happen together and causes issues with overwritting cache.

**Comment by haoadoresorange** [comment]
> There's an important caveat to the use of `prepare` (or `postinstall` for that matter) that's worth mentioning:
> 
> If you have installed `husky` as a devDependency (as is my case and is suggested in the [official docs](https://typicode.github.io/husky/#/?id=install)) and are running `npm publish` as part of a CI or release process or whatever, then it'd simply not work due to `husky not found` error.
> 
> I wouldn't want to resort to simply installing `husky` in the regular dependencies and I know there's a whole [official section](https://typicode.github.io/husky/#/?id=disable-husky-in-cidocker) explaining ways to disable `husky` for specific environments, but I wanted to make people coming across this issue in the future aware of this finding

I second this, the actual best approach even for `yarn` classic is to have it in `postinstall` and use `pinst`. You'll also save yourself sometimes if ever you move to `yarn 2`.

I recommend changing this on the doc.

**Comment by nextglabs** [comment]
> > There's an important caveat to the use of `prepare` (or `postinstall` for that matter) that's worth mentioning:
> > If you have installed `husky` as a devDependency (as is my case and is suggested in the [official docs](https://typicode.github.io/husky/#/?id=install)) and are running `npm publish` as part of a CI or release process or whatever, then it'd simply not work due to `husky not found` error.
> > I wouldn't want to resort to simply installing `husky` in the regular dependencies and I know there's a whole [official section](https://typicode.github.io/husky/#/?id=disable-husky-in-cidocker) explaining ways to disable `husky` for specific environments, but I wanted to make people coming across this issue in the future aware of this finding
> 
> I second this, the actual best approach even for `yarn` classic is to have it in `postinstall` and use `pinst`. You'll also save yourself some times if ever you move to `yarn 2`.
> 
> Maybe you should make this the recommended setup on the doc.

Running `yarn publish --ignore-scripts` seems to solve this issue.
